File: example/BFS/BFS.js
const graph = [[], [2, 3, 8], [1, 7], [1, 4, 5], [3, 5], [3, 4], [7], [2, 6, 8], [1, 7]];
// 각 인덱스에 연결된 노드
const visited = Array(9).fill(false);
let result = "";

bfs(graph, 1, visited);
console.log(result);

function bfs(graph, start, visited) {
  const queue = [start];
  visited[start] = true;
  while (queue.length !== 0) {
    v = queue.shift();
    result += `${v} `;
    for (i of graph[v]) {
      if (!visited[i]) {
        queue.push(i);
        visited[i] = true;
      }
    }
  }
}

File: java/algorithm/BFS.java
import java.util.LinkedList;
import java.util.Queue;

class BFS {
    static void bfs(int start, boolean[] visited, int[][] graph) {
        Queue<Integer> queue = new LinkedList<>();
        queue.add(start);
        visited[start] = true;

        while (!queue.isEmpty()) {
            int v = queue.poll();
            System.out.print(v + " ");

            for (int i = 0; i < graph[v].length; i++) {
                if (!visited[graph[v][i]]) {
                    queue.add(graph[v][i]);
                    visited[graph[v][i]] = true;
                }
            }
        }
    }

    public static void main(String[] args) {
        boolean[] visited = new boolean[7];

        int[][] graph = {
                { 1, 2, 3 },
                { 0, 4, 5 },
                { 0, 6 },
                { 0, 6 },
                { 1 },
                { 1 },
                { 2, 3 },
        };

        bfs(0, visited, graph);
    }
}
